Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.mm

```diff
@@ -0,0 +1,601 @@
+
+    if (!m_firstVideoFrameAudioTime)
+        m_firstVideoFrameAudioTime = sampleTime;
+
+    ASSERT(m_lastEnqueuedRawVideoFrame <= sampleTime);
+    ASSERT(m_lastEnqueuedRawVideoFrame >= m_lastReceivedCompressedVideoFrame);
+    m_lastEnqueuedRawVideoFrame = sampleTime;
+    m_pendingVideoFrames.append({ Ref { frame }, sampleTime });
+    LOG(MediaStream, "appendVideoFrame:enqueuing raw video frame:%f queue:%zu first:%f last:%f", sampleTime.value(), m_pendingVideoFrames.size(), m_pendingVideoFrames.first().second.value(), m_pendingVideoFrames.last().second.value());
+
+    encodePendingVideoFrames();
+}
+
+Seconds MediaRecorderPrivateWriterWebM::nextVideoFrameTime() const
+{
+    ASSERT(!m_hasAudio || m_firstVideoFrameAudioTime);
+    auto now = MonotonicTime::now();
+    auto frameTime = now - m_resumedVideoTime;
+    LOG(MediaStream, "nextVideoFrameTime: frameTime:%f nextVideoFrameTime:%f", frameTime.value(), (frameTime + *m_firstVideoFrameAudioTime).value());
+    frameTime = frameTime + *m_firstVideoFrameAudioTime;
+    // Take the max value from m_lastAudioSampleTime to handle the occasional wall clock time drift and ensure muxed frames times
+    // is always monotonically increasing.
+    return std::max(frameTime, Seconds { m_lastAudioSampleTime.toDouble() });
+}
+
+Seconds MediaRecorderPrivateWriterWebM::resumeVideoTime() const
+{
+    return m_hasAudio ? Seconds { m_resumedAudioTime.toDouble() } : m_currentVideoDuration;
+}
+
+Seconds MediaRecorderPrivateWriterWebM::sampleTime(const RetainPtr<CMSampleBufferRef>& sample) const
+{
+    return Seconds { PAL::CMTimeGetSeconds(PAL::CMSampleBufferGetPresentationTimeStamp(sample.get())) };
+}
+
+void MediaRecorderPrivateWriterWebM::encodePendingVideoFrames()
+{
+    assertIsCurrent(MainThreadDispatcher::singleton());
+
+    if (m_pendingVideoFrames.isEmpty())
+        return;
+
+    m_pendingVideoEncode++;
+
+    m_currentOperation = Ref { m_currentOperation }->whenSettled(MainThreadDispatcher::singleton(), [weakThis = ThreadSafeWeakPtr { *this }, this, pendingVideoFrame = std::exchange(m_pendingVideoFrames, { })]() mutable {
+        RefPtr protectedThis = weakThis.get();
+        if (!protectedThis)
+            return VideoEncoder::EncodePromise::createAndReject("writer deleted"_s);
+        Vector<Ref<VideoEncoder::EncodePromise>> promises { pendingVideoFrame.size() , [&](auto) {
+            auto frame = pendingVideoFrame.takeFirst();
+            bool needVideoKeyframe = false;
+            if (frame.second - m_lastVideoKeyframeTime >= m_maxClusterDuration) {
+                needVideoKeyframe = true;
+                m_lastVideoKeyframeTime = frame.second;
+            }
+            LOG(MediaStream, "encodePendingVideoFrames:encoding video frame:%f", frame.second.value());
+            return m_videoEncoder->encode({ WTFMove(frame.first), frame.second.microsecondsAs<int64_t>(), { } }, needVideoKeyframe);
```

Comment:
It is a bit odd to do Ref{ m_currentOperation } but not protect m_videoEncoder here.

Reply:
m_videoEncoder isn't ref-counted.
protecting m_currentOperation is only there because the new safer CPP requires it (even though it makes no sense here)